File: frontend/lib/realtime.ts
type EventHandler<T> = (payload: T) => void;

export interface MessageEventPayload {
  object: string;
  id: string;
  account_id: string;
  chat_id: string;
  chat_provider_id: string;
  provider_id: string;
  sender_id: string;
  sender_attendee_id: string;
  text: string | null;
  timestamp: string;
  is_sender: number;
  attachments: any[];
  reactions: any[];
  seen: number;
  seen_by: Record<string, any>;
  hidden: number;
  deleted: number;
  edited: number;
  is_event: number;
  delivered: number;
  behavior: number | null;
  original: string | null;
  quoted?: any;
  reply_to?: any;
  event_type?: number | null;
  replies?: number | null;
  reply_by?: string[] | null;
  parent?: string | null;
  subject?: string | null;
  message_type?: string | null;
  attendee_type?: string | null;
  attendee_distance?: number | null;
  sender_urn?: string | null;
  created_at: string;
}

interface RealtimeEvent<T = unknown> {
  type: string;
  payload: T;
}

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";
const WS_PATH = process.env.NEXT_PUBLIC_WS_PATH || "/ws/messages";

function buildWsUrl(): string | null {
  if (typeof window === "undefined") {
    return null;
  }

  try {
    const url = new URL(API_BASE_URL);
    url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
    url.pathname = WS_PATH.startsWith("/") ? WS_PATH : `/${WS_PATH}`;
    url.search = "";
    url.hash = "";
    return url.toString();
  } catch (error) {
    console.error("Invalid NEXT_PUBLIC_API_URL:", error);
    return null;
  }
}

class RealtimeClient {
  private socket: WebSocket | null = null;
  private listeners = new Map<string, Set<EventHandler<any>>>();
  private reconnectAttempts = 0;
  private reconnectTimer: number | null = null;
  private manualClose = false;
  private endpoint: string | null = null;

  connect() {
    if (typeof window === "undefined") return;

    if (!this.endpoint) {
      this.endpoint = buildWsUrl();
    }

    if (!this.endpoint) {
      return;
    }

    if (
      this.socket &&
      (this.socket.readyState === WebSocket.OPEN ||
        this.socket.readyState === WebSocket.CONNECTING)
    ) {
      return;
    }

    this.manualClose = false;
    this.open();
  }

  disconnect() {
    this.manualClose = true;
    if (this.reconnectTimer) {
      window.clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.socket) {
      this.socket.close();
      this.socket = null;
    }
  }

  on<T = unknown>(eventType: string, handler: EventHandler<T>) {
    if (!this.listeners.has(eventType)) {
      this.listeners.set(eventType, new Set());
    }
    this.listeners.get(eventType)!.add(handler as EventHandler<any>);
    this.connect();

    return () => this.off(eventType, handler);
  }

  private off<T>(eventType: string, handler: EventHandler<T>) {
    const handlers = this.listeners.get(eventType);
    if (!handlers) return;
    handlers.delete(handler as EventHandler<any>);
    if (handlers.size === 0) {
      this.listeners.delete(eventType);
    }
  }

  private open() {
    if (!this.endpoint) return;

    this.socket = new WebSocket(this.endpoint);

    this.socket.onopen = () => {
      this.reconnectAttempts = 0;
    };

    this.socket.onmessage = (event) => {
      try {
        const data: RealtimeEvent = JSON.parse(event.data);
        if (!data?.type) {
          return;
        }
        this.emit(data.type, data.payload);
      } catch (error) {
        console.error("Failed to parse realtime payload", error);
      }
    };

    this.socket.onclose = () => {
      this.socket = null;
      if (!this.manualClose) {
        this.scheduleReconnect();
      }
    };

    this.socket.onerror = () => {
      if (this.socket) {
        this.socket.close();
      }
    };
  }

  private scheduleReconnect() {
    if (typeof window === "undefined") return;

    const delay = Math.min(30000, 1000 * 2 ** this.reconnectAttempts);
    this.reconnectAttempts += 1;
    if (this.reconnectTimer) {
      window.clearTimeout(this.reconnectTimer);
    }
    this.reconnectTimer = window.setTimeout(() => {
      this.open();
    }, delay);
  }

  private emit<T>(eventType: string, payload: T) {
    const handlers = this.listeners.get(eventType);
    if (!handlers) return;
    handlers.forEach((handler) => {
      try {
        handler(payload);
      } catch (error) {
        console.error("Realtime handler error", error);
      }
    });
  }
}

export const realtimeClient = new RealtimeClient();

File: backend/app/services/realtime.py
"""Realtime WebSocket utilities for broadcasting message events."""
from __future__ import annotations

import asyncio
import logging
from typing import Any, Dict, Optional, Set

from fastapi import WebSocket
from starlette.websockets import WebSocketState

from app.db.models import MessageModel

logger = logging.getLogger(__name__)


class ConnectionManager:
    """Tracks active WebSocket connections and broadcasts events."""

    def __init__(self) -> None:
        self._connections: Set[WebSocket] = set()
        self._lock = asyncio.Lock()

    async def connect(self, websocket: WebSocket) -> None:
        """Accept and register a WebSocket connection."""
        await websocket.accept()
        async with self._lock:
            self._connections.add(websocket)
        logger.info("WebSocket connected (%s active)", len(self._connections))

    async def disconnect(self, websocket: WebSocket) -> None:
        """Remove a WebSocket connection."""
        async with self._lock:
            self._connections.discard(websocket)
        if websocket.application_state != WebSocketState.DISCONNECTED:
            try:
                await websocket.close()
            except Exception:
                # Ignore failures during cleanup
                pass
        logger.info("WebSocket disconnected (%s active)", len(self._connections))

    async def _snapshot(self) -> Set[WebSocket]:
        """Return a snapshot of current connections."""
        async with self._lock:
            return set(self._connections)

    async def broadcast(self, message: Dict[str, Any]) -> None:
        """Send a JSON message to all active connections."""
        connections = await self._snapshot()
        if not connections:
            return

        stale: Set[WebSocket] = set()
        for connection in connections:
            if connection.application_state != WebSocketState.CONNECTED:
                stale.add(connection)
                continue
            try:
                await connection.send_json(message)
            except Exception as exc:
                logger.warning("WebSocket send failed: %s", exc)
                stale.add(connection)

        for connection in stale:
            await self.disconnect(connection)


manager = ConnectionManager()


def serialize_message(message: MessageModel) -> Dict[str, Any]:
    """Serialize MessageModel into a frontend-friendly dict."""
    return {
        "object": "Message",
        "id": message.id,
        "account_id": message.account_id,
        "chat_id": message.chat_id,
        "chat_provider_id": message.chat_provider_id,
        "provider_id": message.provider_id,
        "sender_id": message.sender_id,
        "sender_attendee_id": message.sender_attendee_id,
        "text": message.text,
        "timestamp": message.timestamp,
        "is_sender": message.is_sender,
        "attachments": message.attachments or [],
        "reactions": message.reactions or [],
        "seen": message.seen,
        "seen_by": message.seen_by or {},
        "hidden": message.hidden,
        "deleted": message.deleted,
        "edited": message.edited,
        "is_event": message.is_event,
        "delivered": message.delivered,
        "sent_by_autopilot": message.sent_by_autopilot,
        "behavior": message.behavior,
        "original": message.original,
        "quoted": message.quoted,
        "reply_to": message.reply_to,
        "event_type": message.event_type,
        "replies": message.replies,
        "reply_by": message.reply_by,
        "parent": message.parent,
        "subject": message.subject,
        "message_type": message.message_type,
        "attendee_type": message.attendee_type,
        "attendee_distance": message.attendee_distance,
        "sender_urn": message.sender_urn,
        "created_at": message.created_at.isoformat(),
    }


def build_message_event(message: MessageModel) -> Dict[str, Any]:
    """Convert a MessageModel into a websocket event envelope."""
    return {
        "type": "message:new",
        "payload": serialize_message(message),
    }


async def broadcast_new_message(message: MessageModel) -> None:
    """Send a `message:new` event to all connected clients."""
    event = build_message_event(message)
    await manager.broadcast(event)


